File: Python/examples/swap.py
# Copyright (C) 2004, 2005, 2006, 2007 StatPro Italia srl
#
# This file is part of QuantLib, a free-software/open-source library
# for financial quantitative analysts and developers - http://quantlib.org/
#
# QuantLib is free software: you can redistribute it and/or modify it under the
# terms of the QuantLib license.  You should have received a copy of the
# license along with this program; if not, please email
# <quantlib-dev@lists.sf.net>. The license is also available online at
# <http://quantlib.org/license.shtml>.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the license for more details.

import QuantLib as ql

# global data
calendar = ql.TARGET()
todaysDate = ql.Date(6, ql.November, 2001)
ql.Settings.instance().evaluationDate = todaysDate
settlementDate = ql.Date(8, ql.November, 2001)

# market quotes
deposits = {
    (1, ql.Weeks): 0.0382,
    (1, ql.Months): 0.0372,
    (3, ql.Months): 0.0363,
    (6, ql.Months): 0.0353,
    (9, ql.Months): 0.0348,
    (1, ql.Years): 0.0345,
}

FRAs = {(3, 6): 0.037125, (6, 9): 0.037125, (9, 12): 0.037125}

futures = {
    ql.Date(19, 12, 2001): 96.2875,
    ql.Date(20, 3, 2002): 96.7875,
    ql.Date(19, 6, 2002): 96.9875,
    ql.Date(18, 9, 2002): 96.6875,
    ql.Date(18, 12, 2002): 96.4875,
    ql.Date(19, 3, 2003): 96.3875,
    ql.Date(18, 6, 2003): 96.2875,
    ql.Date(17, 9, 2003): 96.0875,
}

swaps = {
    (2, ql.Years): 0.037125,
    (3, ql.Years): 0.0398,
    (5, ql.Years): 0.0443,
    (10, ql.Years): 0.05165,
    (15, ql.Years): 0.055175,
}

# convert them to Quote objects
for n, unit in deposits.keys():
    deposits[(n, unit)] = ql.SimpleQuote(deposits[(n, unit)])
for n, m in FRAs.keys():
    FRAs[(n, m)] = ql.SimpleQuote(FRAs[(n, m)])
for d in futures.keys():
    futures[d] = ql.SimpleQuote(futures[d])
for n, unit in swaps.keys():
    swaps[(n, unit)] = ql.SimpleQuote(swaps[(n, unit)])

# build rate helpers

dayCounter = ql.Actual360()
settlementDays = 2
depositHelpers = [
    ql.DepositRateHelper(
        ql.QuoteHandle(deposits[(n, unit)]),
        ql.Period(n, unit),
        settlementDays,
        calendar,
        ql.ModifiedFollowing,
        False,
        dayCounter,
    )
    for n, unit in [(1, ql.Weeks), (1, ql.Months), (3, ql.Months), (6, ql.Months), (9, ql.Months), (1, ql.Years)]
]

dayCounter = ql.Actual360()
settlementDays = 2
fraHelpers = [
    ql.FraRateHelper(
        ql.QuoteHandle(FRAs[(n, m)]), n, m, settlementDays, calendar, ql.ModifiedFollowing, False, dayCounter
    )
    for n, m in FRAs.keys()
]

dayCounter = ql.Actual360()
months = 3
futuresHelpers = [
    ql.FuturesRateHelper(
        ql.QuoteHandle(futures[d]),
        d,
        months,
        calendar,
        ql.ModifiedFollowing,
        True,
        dayCounter,
        ql.QuoteHandle(ql.SimpleQuote(0.0)),
    )
    for d in futures.keys()
]

settlementDays = 2
fixedLegFrequency = ql.Annual
fixedLegTenor = ql.Period(1, ql.Years)
fixedLegAdjustment = ql.Unadjusted
fixedLegDayCounter = ql.Thirty360()
floatingLegFrequency = ql.Semiannual
floatingLegTenor = ql.Period(6, ql.Months)
floatingLegAdjustment = ql.ModifiedFollowing
swapHelpers = [
    ql.SwapRateHelper(
        ql.QuoteHandle(swaps[(n, unit)]),
        ql.Period(n, unit),
        calendar,
        fixedLegFrequency,
        fixedLegAdjustment,
        fixedLegDayCounter,
        ql.Euribor6M(),
    )
    for n, unit in swaps.keys()
]

# term structure handles

discountTermStructure = ql.RelinkableYieldTermStructureHandle()
forecastTermStructure = ql.RelinkableYieldTermStructureHandle()

# term-structure construction

helpers = depositHelpers[:2] + futuresHelpers + swapHelpers[1:]
depoFuturesSwapCurve = ql.PiecewiseFlatForward(settlementDate, helpers, ql.Actual360())

helpers = depositHelpers[:3] + fraHelpers + swapHelpers
depoFraSwapCurve = ql.PiecewiseFlatForward(settlementDate, helpers, ql.Actual360())

# swaps to be priced

swapEngine = ql.DiscountingSwapEngine(discountTermStructure)

nominal = 1000000
length = 5
maturity = calendar.advance(settlementDate, length, ql.Years)
payFixed = True

fixedLegFrequency = ql.Annual
fixedLegAdjustment = ql.Unadjusted
fixedLegDayCounter = ql.Thirty360()
fixedRate = 0.04

floatingLegFrequency = ql.Semiannual
spread = 0.0
fixingDays = 2
index = ql.Euribor6M(forecastTermStructure)
floatingLegAdjustment = ql.ModifiedFollowing
floatingLegDayCounter = index.dayCounter()

fixedSchedule = ql.Schedule(
    settlementDate,
    maturity,
    fixedLegTenor,
    calendar,
    fixedLegAdjustment,
    fixedLegAdjustment,
    ql.DateGeneration.Forward,
    False,
)
floatingSchedule = ql.Schedule(
    settlementDate,
    maturity,
    floatingLegTenor,
    calendar,
    floatingLegAdjustment,
    floatingLegAdjustment,
    ql.DateGeneration.Forward,
    False,
)

spot = ql.VanillaSwap(
    ql.VanillaSwap.Payer,
    nominal,
    fixedSchedule,
    fixedRate,
    fixedLegDayCounter,
    floatingSchedule,
    index,
    spread,
    floatingLegDayCounter,
)
spot.setPricingEngine(swapEngine)

forwardStart = calendar.advance(settlementDate, 1, ql.Years)
forwardEnd = calendar.advance(forwardStart, length, ql.Years)
fixedSchedule = ql.Schedule(
    forwardStart,
    forwardEnd,
    fixedLegTenor,
    calendar,
    fixedLegAdjustment,
    fixedLegAdjustment,
    ql.DateGeneration.Forward,
    False,
)
floatingSchedule = ql.Schedule(
    forwardStart,
    forwardEnd,
    floatingLegTenor,
    calendar,
    floatingLegAdjustment,
    floatingLegAdjustment,
    ql.DateGeneration.Forward,
    False,
)

forward = ql.VanillaSwap(
    ql.VanillaSwap.Payer,
    nominal,
    fixedSchedule,
    fixedRate,
    fixedLegDayCounter,
    floatingSchedule,
    index,
    spread,
    floatingLegDayCounter,
)
forward.setPricingEngine(swapEngine)

# price on the bootstrapped curves


def formatPrice(p, digits=2):
    format = "%%.%df" % digits
    return format % p


def formatRate(r, digits=2):
    format = "%%.%df %%%%" % digits
    return format % (r * 100)


headers = ("term structure", "net present value", "fair spread", "fair fixed rate")
separator = " | "

format = ""
width = 0
for h in headers[:-1]:
    format += "%%%ds" % len(h)
    format += separator
    width += len(h) + len(separator)
format += "%%%ds" % len(headers[-1])
width += len(headers[-1])

rule = "-" * width
dblrule = "=" * width
tab = " " * 8


def report(swap, name):
    print(format % (name, formatPrice(swap.NPV(), 2), formatRate(swap.fairSpread(), 4), formatRate(swap.fairRate(), 4)))


print(dblrule)
print("5-year market swap-rate = %s" % formatRate(swaps[(5, ql.Years)].value()))
print(dblrule)

# price on two different term structures

print(tab + "5-years swap paying %s" % formatRate(fixedRate))
print(separator.join(headers))
print(rule)

discountTermStructure.linkTo(depoFuturesSwapCurve)
forecastTermStructure.linkTo(depoFuturesSwapCurve)
report(spot, "depo-fut-swap")

discountTermStructure.linkTo(depoFraSwapCurve)
forecastTermStructure.linkTo(depoFraSwapCurve)
report(spot, "depo-FRA-swap")

print(rule)

# price the 1-year forward swap

print(tab + "5-years, 1-year forward swap paying %s" % formatRate(fixedRate))
print(rule)

discountTermStructure.linkTo(depoFuturesSwapCurve)
forecastTermStructure.linkTo(depoFuturesSwapCurve)
report(forward, "depo-fut-swap")

discountTermStructure.linkTo(depoFraSwapCurve)
forecastTermStructure.linkTo(depoFraSwapCurve)
report(forward, "depo-FRA-swap")

# modify the 5-years swap rate and reprice

swaps[(5, ql.Years)].setValue(0.046)

print(dblrule)
print("5-year market swap-rate = %s" % formatRate(swaps[(5, ql.Years)].value()))
print(dblrule)

print(tab + "5-years swap paying %s" % formatRate(fixedRate))
print(separator.join(headers))
print(rule)

discountTermStructure.linkTo(depoFuturesSwapCurve)
forecastTermStructure.linkTo(depoFuturesSwapCurve)
report(spot, "depo-fut-swap")

discountTermStructure.linkTo(depoFraSwapCurve)
forecastTermStructure.linkTo(depoFraSwapCurve)
report(spot, "depo-FRA-swap")

print(rule)

print(tab + "5-years, 1-year forward swap paying %s" % formatRate(fixedRate))
print(rule)

discountTermStructure.linkTo(depoFuturesSwapCurve)
forecastTermStructure.linkTo(depoFuturesSwapCurve)
report(forward, "depo-fut-swap")

discountTermStructure.linkTo(depoFraSwapCurve)
forecastTermStructure.linkTo(depoFraSwapCurve)
report(forward, "depo-FRA-swap")

File: R/demo/swap.R

# inspired by python example with the same name

suppressMessages(library(QuantLib))

## global data
calendar <- TARGET()

todaysDate <- DateParser_parseISO("2001-09-06")
invisible(Settings_instance()$setEvaluationDate(d=todaysDate))

settlementDays = 2
settlementDate = Calendar_advance(calendar, todaysDate, settlementDays, "Days")

cat('Today          : ', Date_ISO(todaysDate), "\n")
cat('Settlement Date: ', Date_ISO(settlementDate), "\n")


# ### Market quotes

# %%
deposits = list(
  "3M" = 0.0363
)

# %%
FRAs = list(
  "3x6" = 0.037125,
  "6x9" = 0.037125,
  "9x12" =0.037125
)


# %%
futures = list(
  "2001-12-19" = 96.2875,
  "2002-03-20" = 96.7875,
  "2002-06-19" = 96.9875,
  "2002-09-18" = 96.6875,
  "2002-12-18" = 96.4875,
  "2003-03-19" = 96.3875,
  "2003-06-18" = 96.2875,
  "2003-09-17" = 96.0875
)

# %%
swaps = list(
#  "2Y" = 0.037125,
  "3Y" = 0.0398,
  "5Y" = 0.0443,
  "10Y" = 0.05165,
  "15Y" = 0.055175
  
)

allHelpers_DepoFutSwap = RateHelperVector()
allHelpers_DepoFraSwap = RateHelperVector()

# Rate Helpers - Depos

for (it in names(deposits)) {
  itVal = deposits[[it]]  
  
  dayCounter = Actual360()
  
  dh = DepositRateHelper(
      QuoteHandle(SimpleQuote(itVal)),
      PeriodParser_parse(it),
      settlementDays,
      calendar,
      "ModifiedFollowing",
      F,
      dayCounter
    )  
  
  RateHelperVector_append(allHelpers_DepoFutSwap, dh)
  RateHelperVector_append(allHelpers_DepoFraSwap, dh)
}

# Rate Helpers - FRAs

for (it in names(FRAs)) {
  itVal = FRAs[[it]]  
  
  itSplit = strsplit(x = it, split = "x", fixed = T)[[1]]
  n = as.numeric(itSplit[1])
  m = as.numeric(itSplit[2])
  
  dayCounter = Actual360()
  months = 3

  dh = FraRateHelper(
    QuoteHandle(SimpleQuote(itVal)), n, m, settlementDays, calendar, "ModifiedFollowing", F, dayCounter
  )
  
  RateHelperVector_append(allHelpers_DepoFraSwap, dh)
}
  
  
# Rate Helpers - Futures

for (it in names(futures)) {
  itVal = futures[[it]]  

  dayCounter = Actual360()
  months = 3

  dh = FuturesRateHelper(
    QuoteHandle(SimpleQuote(itVal)),
    DateParser_parseISO(it),
    months,
    calendar,
    "ModifiedFollowing",
    T,
    dayCounter,
    QuoteHandle(SimpleQuote(0.0))    
  )
  
  RateHelperVector_append(allHelpers_DepoFutSwap, dh)
}


# 
# The discount curve for the swaps will come from elsewhere. 
# -> A real application would use some kind of risk-free curve; here we're using a flat one for convenience.
#

# %%
discountTermStructure = YieldTermStructureHandle(FlatForward(settlementDate, 0.04, Actual360()))

# %%

fixedLegFrequency = "Annual"
fixedLegTenor = PeriodParser_parse("1Y") 
fixedLegAdjustment = "Unadjusted"
fixedLegDayCounter = Thirty360(Thirty360_BondBasis_get())
floatingLegFrequency = "Quarterly"
floatingLegTenor = PeriodParser_parse("3M")
floatingLegAdjustment = "ModifiedFollowing"

for (it in names(swaps)) {
  itVal = swaps[[it]]  
  
  dh = SwapRateHelper(
    QuoteHandle(SimpleQuote(itVal)),
    PeriodParser_parse(it),
    calendar,
    fixedLegFrequency,
    fixedLegAdjustment,
    fixedLegDayCounter,
    Euribor3M(),
    QuoteHandle(),
    PeriodParser_parse("0D"),
    discountTermStructure
  )
  
  RateHelperVector_append(allHelpers_DepoFutSwap, dh)
  RateHelperVector_append(allHelpers_DepoFraSwap, dh)
}

# ### Term structure construction

printCurve <- function(curve, helpers) {
  df = NULL
  for (i in seq(1, helpers$size())) {
    h = helpers[i][[1]]
    pillar = RateHelper_pillarDate(h)

    day_counter = Actual365Fixed()
    compounding = "Continuous"
    
    r = YieldTermStructure_zeroRate(curve, pillar, day_counter, compounding, "Annual")
    disc = YieldTermStructure_discount(curve, pillar)
    
    df = rbind(df, data.frame(
      pillar=Date_ISO(pillar), zeroRate=InterestRate_rate(r)*100.0, df=disc, impliedRate=RateHelper_impliedQuote(h)
    ))
  }
  print(df)
}


# %%
forecastTermStructure = RelinkableYieldTermStructureHandle()

# %%
depoFuturesSwapCurve = PiecewiseFlatForward(settlementDate, allHelpers_DepoFutSwap, Actual360())
printCurve(depoFuturesSwapCurve, allHelpers_DepoFutSwap)


# %%
depoFraSwapCurve = PiecewiseFlatForward(settlementDate, allHelpers_DepoFraSwap, Actual360())
printCurve(depoFraSwapCurve, allHelpers_DepoFraSwap)



# ### Swap pricing

# %%
swapEngine = DiscountingSwapEngine(discountTermStructure)

# %%
nominal = 1000000
length = 5
maturity = Calendar_advance(calendar, settlementDate, length, "Years")
payFixed = T

# %%
fixedLegFrequency = "Annual"
fixedLegAdjustment = "Unadjusted"
fixedLegDayCounter = Thirty360(Thirty360_BondBasis_get())
fixedRate = swaps[["5Y"]] # 0.04

# %%
floatingLegFrequency = "Quarterly"
spread = 0.0
fixingDays = 2
index = Euribor3M(forecastTermStructure)
floatingLegAdjustment = "ModifiedFollowing"
floatingLegDayCounter = index$dayCounter()

# %%
fixedSchedule = Schedule(
    settlementDate,
    maturity,
    fixedLegTenor,
    calendar,
    fixedLegAdjustment,
    fixedLegAdjustment,
    DateGeneration_Forward_get(),
    F
)
floatingSchedule = Schedule(
    settlementDate,
    maturity,
    floatingLegTenor,
    calendar,
    floatingLegAdjustment,
    floatingLegAdjustment,
    DateGeneration_Forward_get(),
    F
)

# We'll build a 5-years swap starting spot...

# %%
spot = VanillaSwap(
    Swap_Payer_get(),
    nominal,
    fixedSchedule,
    fixedRate,
    fixedLegDayCounter,
    floatingSchedule,
    index,
    spread,
    floatingLegDayCounter
)
spot$setPricingEngine(swapEngine)



# ...and one starting 1 year forward.

# %%
forwardStart = Calendar_advance(calendar, settlementDate, 1, "Years")
forwardEnd = Calendar_advance(calendar, forwardStart, length, "Years")
fixedSchedule = Schedule(
    forwardStart,
    forwardEnd,
    fixedLegTenor,
    calendar,
    fixedLegAdjustment,
    fixedLegAdjustment,
    DateGeneration_Forward_get(),
    F
)
floatingSchedule = Schedule(
    forwardStart,
    forwardEnd,
    floatingLegTenor,
    calendar,
    floatingLegAdjustment,
    floatingLegAdjustment,
    DateGeneration_Forward_get(),
    F
)

# %%
forward = VanillaSwap(
    Swap_Payer_get(),
    nominal,
    fixedSchedule,
    fixedRate,
    fixedLegDayCounter,
    floatingSchedule,
    index,
    spread,
    floatingLegDayCounter
)
forward$setPricingEngine(swapEngine)


# We'll price them both on the bootstrapped curves.
#
# This is the quoted 5-years market rate; we expect the fair rate of the spot swap to match it.

# %%
print(swaps["5Y"])

showSwap <- function(swap) {
  print(paste("NPV         = ", swap$NPV()))
  print(paste("Fair spread = ", (swap$fairSpread()*100)))
  print(paste("Fair rate   = ", (swap$fairRate()*100)))
}


# %% [markdown]
# These are the results for the 5-years spot swap on the deposit/futures/swap curve...

# %%
forecastTermStructure$linkTo(depoFuturesSwapCurve)
showSwap(spot)

# ...and these are on the deposit/fra/swap curve.

# %%
forecastTermStructure$linkTo(depoFraSwapCurve)
showSwap(spot)

# %% [markdown]
# The same goes for the 1-year forward swap, except for the fair rate not matching the spot rate.

# %%
forecastTermStructure$linkTo(depoFuturesSwapCurve)
showSwap(forward)

# %%
forecastTermStructure$linkTo(depoFraSwapCurve)
showSwap(forward)

